### PR TITLE
As an instructor I want to share a collection group with other bridge instructor users

### PR DESCRIPTION
**Pull request by AndreyLykhoman**
*Description:* Add possibility to share module group with other bridge users.
What was done:
- The new model ContributorPermission
- The views and the forms that related to sharing possibility were added or changed. 
- The class OnlyMyObjectMixin was changed. It checking permission for resources

### CODE REVIEW

File: bridge_adaptivity/module/views.py [hunk 1]
```diff
@@ -183,6 +239,22 @@ class CollectionList(BaseCollectionView, ListView):
     context_object_name = 'collections'
     filter = 'collection_slug'
 
+    def get_context_data(self):
+        context = super().get_context_data()
+        # Get Module Groups where collections are used.
+        mg = ModuleGroup.objects.filter(collections__in=list(context['object_list'])).distinct()
+        # The "name" and "slug" are ModuleGroup fields
+        res = mg.values('name', 'slug', 'collections__slug').filter(
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: module groups related to the user through the owner field and contributor field your current implementation does not look efficient enough. We can discuss it locally.

File: bridge_adaptivity/module/base_views.py [hunk 2]
```diff
@@ -40,9 +40,19 @@ class BaseCollectionView(OnlyMyObjectsMixin, BackURLMixin):
...
 
     fields = ['name', 'slug', 'metadata', 'owner']
+    slug_url_kwarg = 'slug'
+    slug_field = 'slug'
     model = Collection
     ordering = ['id']
 
+    def get_avaliable_resources(self, qs):
+        result_query = (
+            qs.filter(**{self.owner_field: self.request.user}) |
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/base_views.py]: You a right. It seems better to use your variant. 
I rewrote this.

File: bridge_adaptivity/module/base_views.py [hunk 3]
```diff
@@ -40,9 +40,19 @@ class BaseCollectionView(OnlyMyObjectsMixin, BackURLMixin):
     """
 
     fields = ['name', 'slug', 'metadata', 'owner']
+    slug_url_kwarg = 'slug'
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/base_views.py]: These fields are used in this method: https://github.com/harvard-vpal/bridge-adaptivity/pull/230/files#diff-98fb9139d69fa17ef66a927191a2414dR160
'slug_url_kwarg' has the name of kwarg key that used in the URL. Example: 
'group/(?P<group_slug>[\w-]+)/add/collection_order/' -   here it's 'group_slug' but
'collection/(?P<slug>[\w-]+)/' - here it's 'slug'

'slug_field' has the name of a model field. Example: Collection.object.filter(slug='first')

File: bridge_adaptivity/module/forms.py [hunk 4]
```diff
@@ -192,3 +196,49 @@ def save(self, **kwargs):
...
+        """
+        Inner class with metadata options for ContributorPermissionForm.
+        """
+
+        model = ModuleGroup
+        fields = ('contributor_username',)
+
+    def clean(self):
+        super().clean()
+        contributor_username = self.data.get("contributor_username")
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/forms.py]: Yes, I do.  Thanks. 
I fixed it.

File: bridge_adaptivity/module/forms.py [hunk 5]
```diff
@@ -192,3 +196,49 @@ def save(self, **kwargs):
...
+            raise forms.ValidationError(
+                {'contributor_username': "User with this username already has got permission for this resource"}
+            )
+
+        self.cleaned_data["new_consumer_obj"] = new_consumer_obj
+
+        return self.cleaned_data
+
+    def save(self, **kwargs):
+        user = self.cleaned_data.get("new_consumer_obj")
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/forms.py]: Fixed it.

File: bridge_adaptivity/module/mixins/views.py [hunk 6]
```diff
@@ -119,11 +119,15 @@ def get_context_data(self, **kwargs):
 
     def get_form(self):
         form = super().get_form()
-        collections = Collection.objects.filter(
-            owner=self.request.user
+        collections = (
+            Collection.objects.filter(owner=self.request.user) |
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/mixins/views.py]: I fixed it.

File: bridge_adaptivity/module/mixins/views.py [hunk 7]
```diff
@@ -119,11 +119,15 @@ def get_context_data(self, **kwargs):
...
+            Collection.objects.filter(
+                collection_groups__contributors=self.request.user, collection_groups=form.group
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/mixins/views.py]: I changed it.

File: bridge_adaptivity/module/mixins/views.py [hunk 8]
```diff
@@ -137,13 +141,28 @@ def get_form(self):
...
+        return self.get_avaliable_resources(self._filter_resourses(qs))
+
+    def get_avaliable_resources(self, qs):
+        result_query = qs.filter(**{self.owner_field: self.request.user})
+        if self.enable_sharing:
+            result_query = result_query | qs.filter(**{self.contributors_field: self.request.user})
+        return result_query
+
+    def _filter_resourses(self, qs):
+        slug = self.kwargs.get(self.slug_url_kwarg)
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/mixins/views.py]: 'slug_url_kwarg' has the name of kwarg key that used in the URL. Example:
'group/(?P<group_slug>[\w-]+)/add/collection_order/' - here it's 'group_slug'
but
'collection/(?P[\w-]+)/' - here it's 'slug'.

The 'slug' variable contains a slug of a model (Collection or ModuleGroup).

File: bridge_adaptivity/module/models.py [hunk 9]
```diff
@@ -398,6 +401,12 @@ def has_linked_sequences(self):
         return CollectionOrder.objects.filter(group=self, sequence__isnull=False).exists()
 
 
+class ContributorPermission(models.Model):
+    user = models.ForeignKey(BridgeUser, on_delete=models.CASCADE)
+    group = models.ForeignKey(ModuleGroup, on_delete=models.CASCADE)
+    full_permission = models.BooleanField(default=True)
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/models.py]: I think we will add the new fields like 'change_permision', 'read_only_permission', 'share_permission' and etc to this model. And this model will contain all permission settings.

What do you mean? Please, explain it.

File: bridge_adaptivity/module/static/module/js/consumer.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
+$(document).ready(function() {
+
+    $("#contributors .delete-contributor").on("click", function() {
+            var element = $(this);
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/static/module/js/consumer.js]: Done

File: bridge_adaptivity/module/static/module/js/consumer.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+           $.ajax({
+              url: this.dataset["remove_url"],
+            }).done(function() {
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/static/module/js/consumer.js]: I added this case.

File: bridge_adaptivity/module/templates/module/collection_list.html [hunk 12]
```diff
@@ -26,17 +26,19 @@
...
           <td style="max-width: 30%">
-            {% for group in collection.collection_groups.distinct %}
-              <a class="{% if forloop.counter > 10 %}hidden-group {% endif %}label label-info" href="{{ group.get_absolute_url }}">{{ group.name }}</a>
-            {% endfor %}
+            {% for group in avaliable_groups|key:collection.slug %}
+                {% url "module:group-detail"  group.slug as group_detail %}
+                <a class="{% if forloop.counter > 10 %}hidden-group {% endif %}label label-info" href="{{ group_detail }}">{{ group.name }}</a>
+          {% endfor %}
           </td>
-            {% if not request.session.read_only_data %}
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/templates/module/collection_list.html]: I did this accidentally. I fixed it.

File: bridge_adaptivity/module/templates/module/collection_list.html [hunk 13]
```diff
@@ -26,17 +26,19 @@
...
+            {% if not request.session.read_only_data%}
             <td>
               <div>
                     {% url 'module:collection-change' collection.slug as collection_edit %}
                     <a>
                     {% bootstrap_button "Edit" size='sm' icon='edit' extra_classes='modal_launcher' value=collection_edit %}
                     </a>
+                {% if collection.owner == user  %}
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/templates/module/collection_list.html]: Fixed. Thanks.

File: bridge_adaptivity/module/forms.py [hunk 14]
```diff
@@ -192,3 +196,49 @@ def save(self, **kwargs):
...
         self.instance.grading_policy = self.cleaned_data['grading_policy']
         self.instance.save()
+
+
+class ContributorPermissionForm(ModelForm):
+    """
+    Share Module group to Bridge users.
+    """
+
+    contributor_username = forms.CharField()
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/forms.py]: Done

File: bridge_adaptivity/module/templates/module/modulegroup_list.html [hunk 15]
```diff
@@ -47,6 +51,10 @@
                       data-id="{{ group.id }}">
                       {% bootstrap_button "Delete" size='sm' icon='trash' %}
                   </a>
+              {% elif group.owner != user %}
+                <td class="text-right" style="min-width: 90px;">
+                <div>
+                <p>Shared with you by {{ group.owner.username }}</p>
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/templates/module/modulegroup_list.html]: Done

File: bridge_adaptivity/module/tests/test_views.py [hunk 16]
```diff
@@ -528,3 +528,33 @@ def test_list_courses_multiple_sources(
...
+        )
+
+    def test_add_and_delete_contributor(self):
+        url = reverse('module:group-share', kwargs={'group_slug': self.test_cg.slug})
+        data = {
+            "contributor_username": self.contributor.username
+        }
+        response = self.client.post(url, data)
+        self.assertEqual(response.status_code, 200)
+        mg = ModuleGroup.objects.filter(id=self.test_cg.id).first()
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/tests/test_views.py]: I fixed this.

File: bridge_adaptivity/module/tests/test_views.py [hunk 17]
```diff
@@ -528,3 +528,33 @@ def test_list_courses_multiple_sources(
...
+        self.assertTrue(mg.contributors.filter(username=self.contributor.username).exists())
+        url = reverse(
+            'module:group-share-remove',
+            kwargs={'group_slug': self.test_cg.slug, 'username': self.contributor.username}
+        )
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 302)
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/tests/test_views.py]: Done

File: bridge_adaptivity/module/tests/test_models.py [hunk 18]
```diff
@@ -379,3 +379,32 @@ def test_sequence_ui_details(self, option, expected_result):
...
+            password='test_contributor',
+            email='contributor@test.com'
+        )
+
+        self.test_cg = ModuleGroup.objects.create(
+            name='TestColGroup',
+            owner=self.user,
+        )
+
+    def test_create_and_delete_contributor_permission(self):
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/tests/test_models.py]: Done

File: bridge_adaptivity/module/tests/test_views.py [hunk 19]
```diff
@@ -528,3 +528,33 @@ def test_list_courses_multiple_sources(
...
+    @patch('module.tasks.sync_collection_engines.apply_async')
+    def setUp(self, mock_apply_async):
+        super().setUp()
+        self.contributor = BridgeUser.objects.create_user(
+            username='contributor',
+            password='contributor',
+            email='contributor@me.com'
+        )
+
+    def test_add_and_delete_contributor(self):
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/tests/test_views.py]: Done

File: bridge_adaptivity/module/urls.py [hunk 20]
```diff
@@ -16,6 +17,12 @@
     url(r'^group/(?P<group_slug>[\w-]+)/$', ModuleGroupDetail.as_view(), name='group-detail'),
     url(r'^group/(?P<group_slug>[\w-]+)/change/?$', ModuleGroupUpdate.as_view(), name='group-change'),
     url(r'^group/(?P<group_slug>[\w-]+)/delete/?$', ModuleGroupDelete.as_view(), name='group-delete'),
+    url(r"group/(?P<group_slug>[\w-]+)/share/add/$", ModuleGroupShare.as_view(), name="group-share"),
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/urls.py]: Done

File: bridge_adaptivity/module/views.py [hunk 21]
```diff
@@ -169,6 +174,57 @@ def get(self, request, *args, **kwargs):
         return super().get(request, *args, **kwargs)
 
 
+@method_decorator(login_required, name='dispatch')
+class ModuleGroupShare(BaseModuleGroupView, SetUserInFormMixin, ModalFormMixin, UpdateView):
+    template_name = 'module/modals/share_module_group.html'
+    form_class = ContributorPermissionForm
+
+    def get_success_url(self):
+        return self.request.GET.get('return_url') or reverse('module:group-detail', kwargs=self.kwargs)
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: Done

File: bridge_adaptivity/module/templates/module/modulegroup_list.html [hunk 22]
```diff
@@ -47,6 +51,10 @@
                       data-id="{{ group.id }}">
                       {% bootstrap_button "Delete" size='sm' icon='trash' %}
                   </a>
+              {% elif group.owner != user %}
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/templates/module/modulegroup_list.html]: Done

File: bridge_adaptivity/module/views.py [hunk 23]
```diff
@@ -169,6 +174,57 @@ def get(self, request, *args, **kwargs):
...
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        model = context.get('object')
+        if model:
+            context['contributors'] = model.contributors.all()
+        return context
+
+    def form_valid(self, form):
+        if (
+            form.cleaned_data.get('new_consumer_obj') and
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: You are right. I changed this.

File: bridge_adaptivity/module/views.py [hunk 24]
```diff
@@ -169,6 +174,57 @@ def get(self, request, *args, **kwargs):
...
+            form.instance.owner != form.cleaned_data.get('new_consumer_obj')
+        ):
+            super().form_valid(form)
+            form.fields['contributor_username'].help_text = "The new contributor added. You can add another one."
+            form.fields['contributor_username'].initial = None
+            return self.render_to_response(self.get_context_data(form=form, object=form.instance))
+        else:
+            self.form_invalid(form)
+
+    def get_form_kwargs(self):
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: I deleted this.

File: bridge_adaptivity/module/views.py [hunk 25]
```diff
@@ -169,6 +174,57 @@ def get(self, request, *args, **kwargs):
...
+        return result
+
+
+@method_decorator(login_required, name='dispatch')
+class ContributorPermissionDelete(DeleteView):
+    model = ContributorPermission
+
+    def get_success_url(self):
+        return (
+            self.request.GET.get('return_url') or
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: Done

File: bridge_adaptivity/module/views.py [hunk 26]
```diff
@@ -183,6 +239,22 @@ class CollectionList(BaseCollectionView, ListView):
     context_object_name = 'collections'
     filter = 'collection_slug'
 
+    def get_context_data(self):
+        context = super().get_context_data()
+        # Get Module Groups where collections are used.
+        mg = ModuleGroup.objects.filter(collections__in=list(context['object_list'])).distinct()
+        # The "name" and "slug" are ModuleGroup fields
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: Fixed

File: bridge_adaptivity/module/views.py [hunk 27]
```diff
@@ -183,6 +239,22 @@ class CollectionList(BaseCollectionView, ListView):
     context_object_name = 'collections'
     filter = 'collection_slug'
 
+    def get_context_data(self):
+        context = super().get_context_data()
+        # Get Module Groups where collections are used.
+        mg = ModuleGroup.objects.filter(collections__in=list(context['object_list'])).distinct()
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: I added Notes to fix it in the future.

File: bridge_adaptivity/module/views.py (hunk 1 above)
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: I added Notes to fix it in the future.

File: bridge_adaptivity/module/static/module/js/consumer.js [hunk 28]
```diff
@@ -0,0 +1,13 @@
+$(document).ready(function() {
+    $("#contributors .delete-contributor").on("click", function() {
+            var element = $(this);
+           $.ajax({
+              url: this.dataset["remove_url"],
+            }).done(function() {
+              element.closest("tr").remove();
+            }).fail(function (jqXHR, textStatus, errorThrown) {
+              console.log(textStatus + "(" + errorThrown + "): " + jqXHR.responseText.split('\n')[0]);
+              element.closest("tr").children("td.error").append("<p>We fix some problems. Please, try later.</p>");
```
Comment (idegtiarov) [inline bridge_adaptivity/module/static/module/js/consumer.js]: What does this text mean? Who are `We`? And what problems will be fixed? It should be some kind of general message, for example:
```
Cannot be removed, please try again later. If the problem will be procced contact with the Bridge admin.
```

Comment (AndreyLykhoman) [reply bridge_adaptivity/module/static/module/js/consumer.js]: I changed this text.

**Pull request merged**